src/Microsoft.Extensions.ObjectPool/DefaultObjectPool.cs

```diff
@@ -18,43 +22,82 @@ public DefaultObjectPool(IPooledObjectPolicy<T> policy)
 
         public DefaultObjectPool(IPooledObjectPolicy<T> policy, int maximumRetained)
         {
-            if (policy == null)
+            _policy = policy ?? throw new ArgumentNullException(nameof(policy));
+            _isDefaultPolicy = IsDefaultPolicy();
+
+            // -1 due to _firstItem
+            _items = new ObjectWrapper[maximumRetained - 1];
+
+            bool IsDefaultPolicy()
             {
-                throw new ArgumentNullException(nameof(policy));
-            }
+                var type = policy.GetType();
 
-            _policy = policy;
-            _items = new T[maximumRetained];
+                return type.IsGenericType
+                    ? type.GetGenericTypeDefinition() == typeof(DefaultPooledObjectPolicy<>)
```

Comment:
Note: as simple `_isDefaultPolicy = policy is DefaultPooledObjectPolicy<T>;` may no be possible, because this would require the generic constraint `new()`. Adding this constraint to this class is a breaking change, hence this check is used.

Reply:
Sounds good. 

Can you simplify this to `type.IsGenericType && type.GetGenericTypeDefinition()....` the trailing `false` is wierd

Reply:
>  is wierd

Yeah, absolutely weird...shame on me ;-) I'll update the PR with the changes.